specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,141 @@
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `output_word_[n]` 25 columns of the output in words of this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+#### Checks
+
+1. `hash_id` is sequential `next.hash_id - curr.hash_id in [0, 1]`
+2. If we are not in the last round of the circuit or last round of permutation
+   1. `curr.input_len === next.input_len`
+   2. `next.acc_len === curr.acc_len + 136`
+   3. `next.acc_len <= curr.input_len`
+   4. `next.acc_input === curr.acc_input * (r**136) + RLC(input, r)`
+3. `is_last_round_of_circuit or is_end_result === 1`
+   1. Checks the accumulation ends here `assert (curr.input_len - curr.acc_len) in range(136)`
+   2. Clear the variables:  `next.acc_len === 0`, `next.acc_input === 0`
+4. apply to all 25 `output_word_[n]` columns: `next.output_word_[n] === (1 - is_end_result) * curr.output_word_[n]`
+5. `is_end_result === next.hash_id - curr.hash_id`
+
+#### Lookup
+
+To lookup the Keccak256 input to the output, query the following columns:
+
+- `is_end_result`: bool
+- `acc_input`: RLC
+- `output`: RLC
+
+and constrain `is_end_result === 1`
+
+#### Prover behavior
+
+If the prover uses less `Round`s than the circuit provides, the prover should add a dummy hash that has an input size to use all rest of the `Round`s.
+
+### Padding Region
+
+Note that we define a new `acc_len` which increments byte by byte, where the `acc_len` in the lookup region bumps by 136 bytes.
+
+#### Plain behavior: The padding rule
+
+```python
+def get_padding(acc_len) -> bytes:
+    """
+    output big-endian bytes
+    """
+    # note that padlen is at maximum 135
+    padlen = 136 - acc_len 
+    if padlen == 0:
+        # no pad
+        return []
+    elif padlen == 1:
+        # pad 0b10000001
+        return [0x81]
+    elif 1 < padlen < 136:
+        return [0x80] + ([0x00] * int(padlen - 2)) + [0x01]
+    else:
+        raise ValueError("unreachable")
+```
+
+#### Circuit
+
+The Padding Region is a 136-row region.
+
+- `byte` individual byte of the input in big-endian
+- `input_len` Length for correct padding
+- `acc_len` How many bytes we have processed
+- `condition_80_inv` The inverse of `input_len - acc_len` or 0.
+- `padded_byte` Mostly the same as the original `byte` but padded
+- `is_pad_zone` A flag to define the rows that `byte` should be 0
+
+| offset | byte | input_len | acc_len | condition_80_inv | padded_byte | is_pad_zone |
```

Comment:
Padding table is missing accumulator value which is the accomulative RLC.

Reply:
Do we want to do them in this table or should we copy them to another RLC gadget as proposed in point 6? Because to accumulate RLC here we need to place the randomness in this column and complicate the logic.

Reply:
We can have the randomness in an instance column as @han0110 suggested. 
The only issue I see is that we have a lot of rows, which will mean that we have A LOT of instance rows too. And generally we want to have as less PI as possible.

Reply:
We can otherways just copy the randomness once, and copy_constraint it from row to row. So that we make sure that the randomness used is always the same queried from the initial instance column.

Reply:
Added the RLC column for `byte` in [87f13e6](https://github.com/appliedzkp/zkevm-specs/pull/147/commits/87f13e620cd9c159e2d031bba1d844d88c777920)